### PR TITLE
AC-519 kotlin version error, causing build to fail

### PR DESCRIPTION
**Pull request by codepoet2017390**
## Description of what I changed

Updated versions of required Components

## Issue I worked on

JIRA Issue: https://issues.openmrs.org/browse/AC-519

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

**Comment by codepoet2017390** [comment]
@f4ww4z please review

**Comment by codepoet2017390** [comment]
@f4ww4z  I think we need to add multidex

**Comment by ODORA0** [comment]
@codepoet2017390  Can you squash your commits into one? Multiple commits are not advisable https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by ODORA0** [comment]
Did you take note of the travis build failure?

**Comment by f4ww4z** [comment]
@codepoet2017390 Please follow @ODORAO . They are all valid points. As for multidex, you'd need to change your code to meet the normal requirements (without multidex). It is not required to turn it on.

**Comment by codepoet2017390** [comment]
Should i open an Issue to remove those lint errors?

**Comment by f4ww4z** [comment]
> Should i open an Issue to remove those lint errors?

Yes please. Ping me in the new JIRA issue.

File: openmrs-client/build.gradle [hunk 1]
```diff
@@ -99,8 +100,9 @@ repositories {
     maven { url 'http://yanzm.github.io/MaterialTabHost/repository' }
     mavenCentral()
     jcenter()
-    google()
     maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
+    google()
+    maven { url 'https://dl.bintray.com/kotlin/kotlin-dev/' }
```
Comment (f4ww4z) [inline openmrs-client/build.gradle]: @codepoet2017390 just another question. Why is this repository needed?

Comment (codepoet2017390) [reply openmrs-client/build.gradle]: Yes i,m sorry. Its not necessary

**Comment by codepoet2017390** [comment]
@f4ww4z  @ODORA0  i fixed one of the errors reported by lint which was a namespace typo that occurred after incrementing the versions of dependencies and hence Travis passed. Also ive squashed my commits.
Regarding Multidex i'm still unable to build the app without using multidex. Im still trying to find a solution to the multidex issue.

File: openmrs-client/build.gradle (hunk 1 above)
Comment (f4ww4z) [reply openmrs-client/build.gradle]: No problem

**Comment by codepoet2017390** [comment]
@f4ww4z  I read a blog regarding the issue and found that providing support for large range of SDKs which in our app is 16-27 is also a cause of such issue. Upgrading the min SDK version to 24 solved the issue. I am able to successfully build the app.

**Comment by f4ww4z** [comment]
@codepoet2017390 That is not the solution. Android API level 24 (Nougat) is still relatively new. Our app has always been designed to be compatible all the way down to API level 16. Did you try the methods to avoid the 64K limit I sent above?

**Comment by codepoet2017390** [comment]
@f4ww4z  The doc contains 2 methods

1) Review your app's direct and transitive dependencies - As i don't have that good knowledge of all the modules of the project i am still trying to find a way. 

2) Remove unused code with ProGuard - Enabling Proguard shows almost 100 warnings which are listed below

com.google.common.util.concurrent.FuturesGetChecked$GetCheckedTypeValidatorHolder$ClassValueValidator$1: can't find superclass or interface java.lang.ClassValue |  
library class android.content.res.XmlResourceParser extends or implements program class org.xmlpull.v1.XmlPullParser |  
com.google.common.cache.Striped64: can't find referenced class sun.misc.Unsafe |  
com.google.common.cache.Striped64$1: can't find referenced class sun.misc.Unsafe |  
com.google.common.cache.Striped64$Cell: can't find referenced class sun.misc.Unsafe |  
com.google.common.hash.LittleEndianByteArray$UnsafeByteArray: can't find referenced class sun.misc.Unsafe |  
com.google.common.hash.LittleEndianByteArray$UnsafeByteArray$1: can't find referenced class sun.misc.Unsafe |  
com.google.common.hash.LittleEndianByteArray$UnsafeByteArray$2: can't find referenced class sun.misc.Unsafe |  
com.google.common.hash.LittleEndianByteArray$UnsafeByteArray$3: can't find referenced class sun.misc.Unsafe |  
com.google.common.hash.Striped64: can't find referenced class sun.misc.Unsafe |  
com.google.common.hash.Striped64$1: can't find referenced class sun.misc.Unsafe |  
com.google.common.hash.Striped64$Cell: can't find referenced class sun.misc.Unsafe |  
com.google.common.primitives.UnsignedBytes$LexicographicalComparatorHolder$UnsafeComparator: can't find referenced class sun.misc.Unsafe |  
com.google.common.primitives.UnsignedBytes$LexicographicalComparatorHolder$UnsafeComparator$1: can't find referenced class sun.misc.Unsafe |  
com.google.common.util.concurrent.AbstractFuture$UnsafeAtomicHelper: can't find referenced class sun.misc.Unsafe |  
com.google.common.util.concurrent.AbstractFuture$UnsafeAtomicHelper$1: can't find referenced class sun.misc.Unsafe |  
com.google.common.util.concurrent.FuturesGetChecked$GetCheckedTypeValidatorHolder$ClassValueValidator: can't find referenced class java.lang.ClassValue |  
com.google.common.util.concurrent.FuturesGetChecked$GetCheckedTypeValidatorHolder$ClassValueValidator$1: can't find referenced class java.lang.ClassValue |  
com.google.errorprone.annotations.ForOverride: can't find referenced class javax.lang.model.element.Modifier |  
com.google.errorprone.annotations.IncompatibleModifiers: can't find referenced class javax.lang.model.element.Modifier |  
com.google.errorprone.annotations.RequiredModifiers: can't find referenced class javax.lang.model.element.Modifier |  
com.google.errorprone.annotations.Var: can't find referenced class javax.lang.model.element.Modifier |  
org.joda.time.DateMidnight: can't find referenced class org.joda.convert.FromString |  
org.joda.time.DateTime: can't find referenced class org.joda.convert.FromString |  
org.joda.time.DateTimeZone: can't find referenced class org.joda.convert.FromString |  
org.joda.time.DateTimeZone: can't find referenced class org.joda.convert.ToString |  
org.joda.time.Days: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Days: can't find referenced class org.joda.convert.ToString |  
org.joda.time.Duration: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Hours: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Hours: can't find referenced class org.joda.convert.ToString |  
org.joda.time.Instant: can't find referenced class org.joda.convert.FromString |  
org.joda.time.LocalDate: can't find referenced class org.joda.convert.FromString |  
org.joda.time.LocalDate: can't find referenced class org.joda.convert.ToString |  
org.joda.time.LocalDateTime: can't find referenced class org.joda.convert.FromString |  
org.joda.time.LocalDateTime: can't find referenced class org.joda.convert.ToString |  
org.joda.time.LocalTime: can't find referenced class org.joda.convert.FromString |  
org.joda.time.LocalTime: can't find referenced class org.joda.convert.ToString |  
org.joda.time.Minutes: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Minutes: can't find referenced class org.joda.convert.ToString |  
org.joda.time.MonthDay: can't find referenced class org.joda.convert.FromString |  
org.joda.time.MonthDay: can't find referenced class org.joda.convert.ToString |  
org.joda.time.Months: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Months: can't find referenced class org.joda.convert.ToString |  
org.joda.time.MutableDateTime: can't find referenced class org.joda.convert.FromString |  
org.joda.time.MutableDateTime: can't find referenced class org.joda.convert.ToString |  
org.joda.time.MutablePeriod: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Period: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Seconds: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Seconds: can't find referenced class org.joda.convert.ToString |  
org.joda.time.Weeks: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Weeks: can't find referenced class org.joda.convert.ToString |  
org.joda.time.YearMonth: can't find referenced class org.joda.convert.FromString |  
org.joda.time.YearMonth: can't find referenced class org.joda.convert.ToString |  
org.joda.time.Years: can't find referenced class org.joda.convert.FromString |  
org.joda.time.Years: can't find referenced class org.joda.convert.ToString |  
org.joda.time.base.AbstractDuration: can't find referenced class org.joda.convert.ToString |  
org.joda.time.base.AbstractInstant: can't find referenced class org.joda.convert.ToString |  
org.joda.time.base.AbstractPeriod: can't find referenced class org.joda.convert.ToString |  
org.slf4j.LoggerFactory: can't find referenced class org.slf4j.impl.StaticLoggerBinder |  
org.slf4j.MDC: can't find referenced class org.slf4j.impl.StaticMDCBinder |  
org.slf4j.MarkerFactory: can't find referenced class org.slf4j.impl.StaticMarkerBinder |  
rx.internal.util.unsafe.BaseLinkedQueueConsumerNodeRef: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.BaseLinkedQueueProducerNodeRef: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.ConcurrentCircularArrayQueue: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.ConcurrentSequencedCircularArrayQueue: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.MpmcArrayQueueConsumerField: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.MpmcArrayQueueProducerField: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.MpscLinkedQueue: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.SpmcArrayQueueConsumerField: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.SpmcArrayQueueProducerField: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.SpscArrayQueue: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.SpscUnboundedArrayQueue: can't find referenced class sun.misc.Unsafe |  
rx.internal.util.unsafe.UnsafeAccess: can't find referenced class sun.misc.Unsafe |  
library class android.content.Intent depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.content.IntentFilter depends on program class org.xmlpull.v1.XmlSerializer |  
library class android.content.IntentFilter depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.content.res.ColorStateList depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.AnimatedVectorDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.AnimationDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.BitmapDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.ClipDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.ColorDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.Drawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.DrawableWrapper depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.GradientDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.InsetDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.LayerDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.RippleDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.ScaleDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.ShapeDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.StateListDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.graphics.drawable.VectorDrawable depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.util.Xml depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.util.Xml depends on program class org.xmlpull.v1.XmlSerializer |  
library class android.util.Xml depends on program class org.xmlpull.v1.XmlPullParser |  
library class android.view.LayoutInflater depends on program class org.xmlpull.v1.XmlPullParser |  
there were 179 unresolved references to classes or interfaces. |  
there were 32 instances of library classes depending on program classes. |  
Exception while processing task java.io.IOException: Please correct the above warnings first. |  

Actually i haven't come across the dex limit error ever before so please help me proceed.

**Comment by f4ww4z** [comment]
Hi, did you check out https://stackoverflow.com/questions/33865425/getting-warnings-and-errors-after-enabling-proguard ? You are heading in the right direction. Do try to research more on any problems you face.

**Comment by codepoet2017390** [comment]
@f4ww4z Just had a doubt, proguard should be used for release builds as mentioned on the official android developers website but presently even our debug code is crossing the dex limit and thus i have to use proguard for debug build also.
So is this right or should i change something?

**Comment by f4ww4z** [comment]
@codepoet2017390 Yes, it is very acceptable. Add [this to the](https://stackoverflow.com/a/39141843/8434188) gradle file.

**Comment by codepoet2017390** [comment]
@f4ww4z Already added and trying to implement the custom rules in proguard-rules.pro files :)

**Comment by f4ww4z** [comment]
@codepoet2017390 I finally agreed to enable multidex as it doesn't cost any noticeable performance issues and that the codebase should be modular with many methods. Please enable that and make sure travis build passes.

Did you also get proguard to work?

**Comment by codepoet2017390** [comment]
Thanks @f4ww4z  i tried my best to find a workaround but ultimately Multidex would come out as a solution at one stage or another. Also mutlidex was required for implementing Room. Regarding
 Pro-guard its already implemented for production builds and I will implement it for debug builds if you ask to?

**Comment by f4ww4z** [comment]
@codepoet2017390 minify is not needed for debug builds where faster build speed is most wanted by developers.

**Comment by codepoet2017390** [comment]
@f4ww4z Ok so I'll implement proguard for debug builds and Make minify false 
Anything else to be changed?

**Comment by f4ww4z** [comment]
@codepoet2017390 correction : don't change the `debug` config and just enable minify for `release` builds. Then commit and I'll do a final review.

**Comment by codepoet2017390** [comment]
@f4ww4z  please review

**Comment by f4ww4z** [comment]
@codepoet2017390 Next time please rename your commit to start with the issue number.

**Pull request merged**